File: saksflyt/src/main/java/no/nav/melosys/saksflyt/steg/medl/LagreMedlemsperiodeMedl.java
package no.nav.melosys.saksflyt.steg.medl;

import no.nav.melosys.domain.Medlemskapsperiode;
import no.nav.melosys.domain.saksflyt.ProsessSteg;
import no.nav.melosys.domain.saksflyt.Prosessinstans;
import no.nav.melosys.saksflyt.steg.StegBehandler;
import no.nav.melosys.service.behandling.BehandlingsresultatService;
import no.nav.melosys.service.medlemskapsperiode.MedlemskapsperiodeService;
import org.springframework.stereotype.Component;

@Component
public class LagreMedlemsperiodeMedl implements StegBehandler {

    private final MedlemskapsperiodeService medlemskapsperiodeService;
    private final BehandlingsresultatService behandlingsresultatService;

    public LagreMedlemsperiodeMedl(MedlemskapsperiodeService medlemskapsperiodeService,
                                   BehandlingsresultatService behandlingsresultatService) {
        this.medlemskapsperiodeService = medlemskapsperiodeService;
        this.behandlingsresultatService = behandlingsresultatService;
    }

    @Override
    public ProsessSteg inngangsSteg() {
        return ProsessSteg.LAGRE_MEDLEMSKAPSPERIODE_MEDL;
    }

    @Override
    public void utfør(Prosessinstans prosessinstans) {
        var behandling = prosessinstans.getBehandling();
        long behandlingId = prosessinstans.getBehandling().getId();
        var behandlingsresultat = behandlingsresultatService.hentBehandlingsresultat(behandlingId);

        var innvilgedeMedlemskapsperioder = behandlingsresultat.finnMedlemskapsperioder()
            .stream().filter(Medlemskapsperiode::erInnvilget).toList();
        if (behandling.erNyVurdering() && behandling.getOpprinneligBehandling() != null) {
            medlemskapsperiodeService.erstattMedlemskapsperioder(innvilgedeMedlemskapsperioder, behandling.getOpprinneligBehandling().getId(), behandlingId);
        } else {
            for (Medlemskapsperiode medlemskapsperiode : innvilgedeMedlemskapsperioder) {
                medlemskapsperiodeService.opprettEllerOppdaterMedlPeriode(behandlingId, medlemskapsperiode);
            }
        }
    }
}

File: saksflyt/src/main/kotlin/no/nav/melosys/saksflyt/steg/medl/LagreMedlemsperiodeMedl.kt
package no.nav.melosys.saksflyt.steg.medl

import mu.KotlinLogging
import no.nav.melosys.domain.Medlemskapsperiode
import no.nav.melosys.saksflyt.steg.StegBehandler
import no.nav.melosys.saksflytapi.domain.ProsessSteg
import no.nav.melosys.saksflytapi.domain.Prosessinstans
import no.nav.melosys.service.behandling.BehandlingsresultatService
import no.nav.melosys.service.ftrl.medlemskapsperiode.MedlemskapsperiodeService
import org.springframework.stereotype.Component

private val log = KotlinLogging.logger { }

@Component
class LagreMedlemsperiodeMedl(
    private val medlemskapsperiodeService: MedlemskapsperiodeService,
    private val behandlingsresultatService: BehandlingsresultatService
) : StegBehandler {
    override fun inngangsSteg(): ProsessSteg {
        return ProsessSteg.LAGRE_MEDLEMSKAPSPERIODE_MEDL
    }

    override fun utfør(prosessinstans: Prosessinstans) {
        val behandling = prosessinstans.hentBehandling
        val behandlingID = prosessinstans.hentBehandling.id
        log.info { "[STEP-START] LagreMedlemsperiodeMedl behandlingId=$behandlingID" }
        val behandlingsresultat = behandlingsresultatService.hentBehandlingsresultat(behandlingID)
        log.info { "[STEP-LOADED] LagreMedlemsperiodeMedl behandlingId=$behandlingID type=${behandlingsresultat.type}" }

        if(behandling.erEøsPensjonist()){
            return
        }

        if (behandling.erAndregangsbehandling() && behandling.opprinneligBehandling != null) {
            medlemskapsperiodeService.erstattMedlemskapsperioder(
                behandlingID,
                behandling.opprinneligBehandling!!.id,
                behandlingsresultat.medlemskapsperioder.toList()
            )
        } else {
            behandlingsresultat.medlemskapsperioder.filter(Medlemskapsperiode::erInnvilget).forEach {
                medlemskapsperiodeService.opprettEllerOppdaterMedlPeriode(behandlingID, it)
            }
        }
    }
}
